pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1385,4 +1386,89 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
       }
     }
   }
+
+  /**
+   * Creates PROPERTYSTORE and IDEALSTATE entries for each partitionGroup to resume realtime table consumption
+   * @param tableName
+   */
+  public void resumeRealtimeTableConsumption(String tableName) {
+    TableConfig tableConfig = _helixResourceManager.getTableConfig(tableName, TableType.REALTIME);
+    String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(tableName);
+
+    InstancePartitions instancePartitions =
+        InstancePartitionsUtils.fetchOrComputeInstancePartitions(_helixResourceManager.getHelixZkManager(), tableConfig,
+            InstancePartitionsType.CONSUMING);
+    Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap =
+        Collections.singletonMap(InstancePartitionsType.CONSUMING, instancePartitions);
+    IdealState idealState =
+        HelixHelper.getTableIdealState(_helixResourceManager.getHelixZkManager(), realtimeTableName);
+    int numReplicas = getNumReplicas(tableConfig, instancePartitions);
+    Map<String, Map<String, String>> instanceStatesMap = idealState.getRecord().getMapFields();
+
+    PartitionLevelStreamConfig streamConfig = new PartitionLevelStreamConfig(tableConfig.getTableName(),
+        IngestionConfigUtils.getStreamConfigMap(tableConfig));
+    List<PartitionGroupConsumptionStatus> currentPartitionGroupConsumptionStatusList =
+        getPartitionGroupConsumptionStatusList(idealState, streamConfig);
+    // Read the smallest offset when a new partition is detected
+    OffsetCriteria originalOffsetCriteria = streamConfig.getOffsetCriteria();
+    streamConfig.setOffsetCriteria(OffsetCriteria.SMALLEST_OFFSET_CRITERIA);
+    List<PartitionGroupMetadata> newPartitionGroupMetadataList =
+        getNewPartitionGroupMetadataList(streamConfig, currentPartitionGroupConsumptionStatusList);
+    streamConfig.setOffsetCriteria(originalOffsetCriteria);
+
+    int numPartitionGroups = newPartitionGroupMetadataList.size();
+    Map<Integer, SegmentZKMetadata> partitionToLatestSegment = getLatestSegmentZKMetadataMap(realtimeTableName);
+    for (PartitionGroupMetadata partitionGroupMetadata : newPartitionGroupMetadataList) {
```

Comment:
Intentional, yes. But we can discuss this offline. Idea was to keep this as a generic resumeConsumption API, which also happens to double up as a fix for deleted segments.

Reply:
I see. But it's not going to be a generic resume consumption Api unless all the other cases from ensureAllPartitions are also included. Now we have two ways to resume consumption, both doing only partial cases.

Reply:
What do you think if we make the following changes to `ensureAllPartitionsConsuming` method :
```java
ensureAllPartitionsConsuming(..., boolean enforcedByAdmin) {
  ...
  if (enforcedByAdmin) {
    // handle the case where a realtime segment is manually deleted 
  }
  ...
}
```
Then this method is called in the resume endpoint with enforcedByAdmin = true and in the periodic task with enforcedByAdmin = false.
Later if we have more cases that require admin intervention, we can add them to the `if (enforcedByAdmin) {...}` block.

Reply:
this is a great idea. +1 on doing it this way

Reply:
Ack. will move this logic into `ensureAllPartitionsConsuming`
In that case, we'd be creating segments for new partition groups as well @npawar ?

Reply:
imo this is okay, as we have all logic in one place and we're not just creating new partition groups but running the whole flow with all the cases. My main concern was that we were just doing half of the things

Reply:
Ack. This question turned out to be redundant since creating new consuming segments for new partitions is already handled by `ensureAllPartitionsConsuming`. 
All in all, the API does everything `ensureAllPartitionsConsuming` already does, and additionally also creates consuming segments for partitions where CONSUMING segments were deleted.

Reply:
IMO, having a single flag of `enforcedByAdmin` can be quite confusing and hard to use because this module (segment manager) should not be aware of the intention of the admin. I'd suggest making the flag specific and self-explained, such as `recreateDeletedConsumingSegment`. In the future if we want to add other checks, we can introduce other flags, or wrap the flags into a options class, but each flag should be specific. @sajjad-moradi @npawar What's your opinion on this suggestion?

Reply:
my thought was also of a generic name like "fixAllCases" or "fixManualErrors" . But i see your point, self-explanatory name would be a good start and introduce more as and when needed.

Reply:
That's a valid point. We can start with this specific flag now and later on, when we add more cases, we rename the flag to be more generic or add more options.

Reply:
Ack. Renaming the flag. Wrapping these flags into a Config / Props class eventually sounds good to me.